In [2]:
import pandas as pd # library for data analsysis
import numpy as np

!conda install -c conda-forge geopy --yes 
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests
import json # library to handle JSON files
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

# for webscraping import Beautiful Soup 
from bs4 import BeautifulSoup

import xml

!conda install -c conda-forge folium=0.5.0 --yes
import folium # map rendering library

print('Libraries imported.')

Solving environment: done

## Package Plan ##

  environment location: /opt/conda/envs/Python36

  added / updated specs: 
    - geopy


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    openssl-1.1.1d             |       h516909a_0         2.1 MB  conda-forge
    geographiclib-1.50         |             py_0          34 KB  conda-forge
    python_abi-3.6             |          1_cp36m           4 KB  conda-forge
    ca-certificates-2019.11.28 |       hecc5488_0         145 KB  conda-forge
    geopy-1.21.0               |             py_0          58 KB  conda-forge
    certifi-2019.11.28         |   py36h9f0ad1d_1         149 KB  conda-forge
    ------------------------------------------------------------
                                           Total:         2.5 MB

The following NEW packages will be INSTALLED:

    geographiclib:   1.50-py_0         conda-forge
    geopy:           1

In [8]:
r = requests.get('https://usc.data.socrata.com/api/geospatial/r8qd-yxsr?method=export&format=GeoJSON')
losangeles_data = r.json()

In [9]:
neighborhoods_data = losangeles_data['features']

In [12]:
neighborhoods_data[2]

{'type': 'Feature',
 'properties': {'external_i': 'agoura-hills',
  'name': 'Agoura Hills',
  'location': 'POINT(34.146736499122795 -118.75988450000015)',
  'latitude': '-118.75988450000015',
  'slug_1': None,
  'sqmi': '8.14676029818',
  'display_na': 'Agoura Hills L.A. County Neighborhood (Current)',
  'set': 'L.A. County Neighborhoods (Current)',
  'slug': 'agoura-hills',
  'longitude': '34.146736499122795',
  'name_1': None,
  'kind': 'L.A. County Neighborhood (Current)',
  'type': 'standalone-city'},
 'geometry': {'type': 'MultiPolygon',
  'coordinates': [[[[-118.76192500000009, 34.16820299912263],
     [-118.72631699999998, 34.16787499912258],
     [-118.73232400000018, 34.1543019991227],
     [-118.73706000000027, 34.15435299912277],
     [-118.73702385723149, 34.14304324369543],
     [-118.73587193294368, 34.14267020794711],
     [-118.73493704578691, 34.14215345623102],
     [-118.7306383000593, 34.139239215941366],
     [-118.72918210881832, 34.13843288950915],
     [-118.727

In [13]:
neighborhoods = pd.DataFrame(columns=['Neighborhood', 'Latitude', 'Longitude'] )
neighborhoods

Empty DataFrame
Columns: [Neighborhood, Latitude, Longitude]
Index: []

In [14]:
for data in neighborhoods_data:
    neighborhood_name = data['properties']['name']
    neighborhood_lat = data['properties']['latitude']
    neighborhood_lon = data['properties']['longitude']
    
    neighborhoods = neighborhoods.append({'Neighborhood': neighborhood_name,
                                          'Latitude': neighborhood_lat,
                                          'Longitude': neighborhood_lon}, ignore_index=True)

In [16]:
neighborhoods.head()

Neighborhood             Latitude           Longitude
0            Acton  -118.16981019229348  34.497355239240846
1  Adams-Normandie  -118.30020800000011  34.031461499124156
2     Agoura Hills  -118.75988450000015  34.146736499122795
3       Agua Dulce   -118.3171036690717  34.504926999796837
4         Alhambra  -118.13651200000021  34.085538999123571

In [17]:
address = 'Los Angeles, CA'

geolocator = Nominatim()
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Los Angeles, CA are {}, {}.'.format(latitude, longitude))

/opt/conda/envs/Python36/lib/python3.6/site-packages/ipykernel/__main__.py:3: DeprecationWarning: Using Nominatim with the default "geopy/1.18.1" `user_agent` is strongly discouraged, as it violates Nominatim's ToS https://operations.osmfoundation.org/policies/nominatim/ and may possibly cause 403 and 429 HTTP errors. Please specify a custom `user_agent` with `Nominatim(user_agent="my-application")` or by overriding the default `user_agent`: `geopy.geocoders.options.default_user_agent = "my-application"`. In geopy 2.0 this will become an exception.
  app.launch_new_instance()


The geograpical coordinate of Los Angeles, CA are 34.0536909, -118.2427666.


## The values in the JSON file are all around 34, -118. However, latitude and longitude are reversed, so we will change the column names to reflect this.

In [31]:
neighborhoods.rename(columns = {'Latitude':'Longitude', 'Longitude':'Latitude'}, inplace=True)
neighborhoods.head()

Neighborhood            Longitude            Latitude
0            Acton  -118.16981019229348  34.497355239240846
1  Adams-Normandie  -118.30020800000011  34.031461499124156
2     Agoura Hills  -118.75988450000015  34.146736499122795
3       Agua Dulce   -118.3171036690717  34.504926999796837
4         Alhambra  -118.13651200000021  34.085538999123571

In [32]:
neighborhoods.shape

(272, 3)

In [33]:
neighborhoods.dropna(subset=['Latitude'], inplace=True)
neighborhoods.dropna(subset=['Longitude'], inplace=True)
neighborhoods.shape

(272, 3)

In [24]:
map_los_angeles = folium.Map(location=[latitude, longitude], zoom_start=10)
map_los_angeles

In [34]:
for lat, lng, in zip(neighborhoods['Latitude'], neighborhoods['Longitude']):
    map_los_angeles.add_child(
        folium.features.CircleMarker(
            [float(lat), float(lng)],
            radius=5, # define how big you want the circle markers to be
            color='yellow',
            fill=True,
            fill_color='blue',
            fill_opacity=0.6
        )
    )
    
map_los_angeles

# Foursquare API

In [35]:
CLIENT_ID = 'ERFCS0HI2ODMI5PVVGK2FXA0BBDO5245K3YHLDTXF4UBWPBG' # your Foursquare ID
CLIENT_SECRET = 'RP2Z2TV0DWKA10HHRQI5YYI5LCVBYTWTQHNMQ41PXAN2BRM0' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: ERFCS0HI2ODMI5PVVGK2FXA0BBDO5245K3YHLDTXF4UBWPBG
CLIENT_SECRET:RP2Z2TV0DWKA10HHRQI5YYI5LCVBYTWTQHNMQ41PXAN2BRM0


In [36]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [38]:
LIMIT=100

los_angeles_venues = getNearbyVenues(names=neighborhoods['Neighborhood'],
                                   latitudes=neighborhoods['Latitude'],
                                   longitudes=neighborhoods['Longitude']
                                  )

Acton
Adams-Normandie
Agoura Hills
Agua Dulce
Alhambra
Alondra Park
Artesia
Altadena
Angeles Crest
Arcadia
Arleta
Arlington Heights
Athens
Atwater Village
Avalon
Avocado Heights
Azusa
Vermont-Slauson
Baldwin Hills/Crenshaw
Baldwin Park
Bel-Air
Bellflower
Bell Gardens
Green Valley
Bell
Beverly Crest
Beverly Grove
Burbank
Koreatown
Beverly Hills
Beverlywood
Boyle Heights
Bradbury
Brentwood
Broadway-Manchester
Calabasas
Canoga Park
Carson
Carthay
Castaic Canyons
Chatsworth
Castaic
Central-Alameda
Century City
Cerritos
Charter Oak
Chatsworth Reservoir
Chesterfield Square
Cheviot Hills
Chinatown
Citrus
Claremont
Northridge
Commerce
Compton
Cypress Park
La Mirada
Covina
Cudahy
Culver City
Del Aire
Del Rey
Desert View Highlands
Diamond Bar
Downey
Downtown
Duarte
Eagle Rock
East Compton
East Hollywood
East La Mirada
Elizabeth Lake
East Los Angeles
East Pasadena
East San Gabriel
Echo Park
El Monte
El Segundo
El Sereno
Elysian Park
Elysian Valley
Vermont Square
Encino
Exposition Park
Fairfax
Flo

In [39]:
print(los_angeles_venues.shape)
los_angeles_venues.head()

(2989, 7)


Neighborhood Neighborhood Latitude Neighborhood Longitude  \
0            Acton    34.497355239240846    -118.16981019229348   
1            Acton    34.497355239240846    -118.16981019229348   
2  Adams-Normandie    34.031461499124156    -118.30020800000011   
3  Adams-Normandie    34.031461499124156    -118.30020800000011   
4  Adams-Normandie    34.031461499124156    -118.30020800000011   

                Venue  Venue Latitude  Venue Longitude  \
0    Epik Engineering       34.498718      -118.168046   
1  Alma Gardening Co.       34.494762      -118.172550   
2   Orange Door Sushi       34.032485      -118.299368   
3               Shell       34.033095      -118.300025   
4       Sushi Delight       34.032445      -118.299525   

               Venue Category  
0  Construction & Landscaping  
1  Construction & Landscaping  
2            Sushi Restaurant  
3                 Gas Station  
4            Sushi Restaurant

In [40]:
los_angeles_venues.groupby('Neighborhood').count()

Neighborhood Latitude  Neighborhood Longitude  \
Neighborhood                                                              
Acton                                         2                       2   
Adams-Normandie                               9                       9   
Agoura Hills                                 30                      30   
Agua Dulce                                    1                       1   
Alhambra                                     13                      13   
Alondra Park                                  4                       4   
Altadena                                      5                       5   
Arcadia                                      21                      21   
Arleta                                        4                       4   
Arlington Heights                            27                      27   
Artesia                                      31                      31   
Atwater Village                              22                      22   
Avalon                                        4                       4   
Avocado Heights                              12                      12   
Azusa                                         9                       9   
Baldwin Hills/Crenshaw                        3                       3   
Baldwin Park                                  7                       7   
Bel-Air                                       3                       3   
Bell                                          1                       1   
Bell Gardens                                 12                      12   
Bellflower                                   15                      15   
Beverly Crest                                 1                       1   
Beverly Grove                               100                     100   
Beverly Hills                                 3                       3   
Beverlywood                                   5                       5   
Boyle Heights                                 8                       8   
Bradbury                                      2                       2   
Brentwood                                     2                       2   
Broadway-Manchester                           5                       5   
Burbank                                      18                      18   
...                                         ...                     ...   
Vermont Square                                6                       6   
Vermont Vista                                 1                       1   
Vermont-Slauson                               8                       8   
Vernon                                        2                       2   
Veterans Administration                       4                       4   
Vincent                                       1                       1   
Walnut                                        2                       2   
Walnut Park                                   7                       7   
Watts                                         7                       7   
West Adams                                   11                      11   
West Carson                                   4                       4   
West Compton                                  2                       2   
West Covina                                   6                       6   
West Hills                                    7                       7   
West Hollywood                               65                      65   
West Los Angeles                             31                      31   
West Puente Valley                            8                       8   
West San Dimas                                2                       2   
West Whittier-Los Nietos                      3                       3   
Westchester                                  47                      47   
Westlake                                     13    

In [41]:
print('There are {} uniques categories.'.format(len(los_angeles_venues['Venue Category'].unique())))

There are 322 uniques categories.


# Analyzing the Data

#### Frequency of each venue category.
We use one-hot encoding for this.

In [42]:
los_angeles_onehot = pd.get_dummies(los_angeles_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
los_angeles_onehot['Neighborhood'] = los_angeles_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [los_angeles_onehot.columns[-1]] + list(los_angeles_onehot.columns[:-1])
los_angeles_onehot = los_angeles_onehot[fixed_columns]

los_angeles_onehot.head()

Yoga Studio  ATM  Accessories Store  Airport  Alternative Healer  \
0            0    0                  0        0                   0   
1            0    0                  0        0                   0   
2            0    0                  0        0                   0   
3            0    0                  0        0                   0   
4            0    0                  0        0                   0   

   American Restaurant  Antique Shop  Arcade  Argentinian Restaurant  \
0                    0             0       0                       0   
1                    0             0       0                       0   
2                    0             0       0                       0   
3                    0             0       0                       0   
4                    0             0       0                       0   

   Art Gallery  ...  Video Store  Vietnamese Restaurant  Warehouse Store  \
0            0  ...            0                      0                0   
1            0  ...            0                      0                0   
2            0  ...            0                      0                0   
3            0  ...            0                      0                0   
4            0  ...            0                      0                0   

   Watch Shop  Water Park  Weight Loss Center  Wine Bar  Wine Shop  \
0           0           0                   0         0          0   
1           0           0                   0         0          0   
2           0           0                   0         0          0   
3           0           0                   0         0          0   
4           0           0                   0         0          0   

   Wings Joint  Women's Store  
0            0              0  
1            0              0  
2            0              0  
3            0              0  
4            0              0  

[5 rows x 322 columns]

In [43]:
los_angeles_onehot.shape

(2989, 322)

#### Grouping by neighborhood.
We calculate the mean of each frequency.

In [45]:
los_angeles_grouped = los_angeles_onehot.groupby('Neighborhood').mean().reset_index()
los_angeles_grouped.head()

Neighborhood  Yoga Studio  ATM  Accessories Store  Airport  \
0            Acton          0.0  0.0                0.0      0.0   
1  Adams-Normandie          0.0  0.0                0.0      0.0   
2     Agoura Hills          0.0  0.0                0.0      0.0   
3       Agua Dulce          0.0  0.0                0.0      1.0   
4         Alhambra          0.0  0.0                0.0      0.0   

   Alternative Healer  American Restaurant  Antique Shop  Arcade  \
0                 0.0             0.000000           0.0     0.0   
1                 0.0             0.000000           0.0     0.0   
2                 0.0             0.033333           0.0     0.0   
3                 0.0             0.000000           0.0     0.0   
4                 0.0             0.000000           0.0     0.0   

   Argentinian Restaurant  ...  Video Store  Vietnamese Restaurant  \
0                     0.0  ...     0.000000                    0.0   
1                     0.0  ...     0.000000                    0.0   
2                     0.0  ...     0.000000                    0.0   
3                     0.0  ...     0.000000                    0.0   
4                     0.0  ...     0.076923                    0.0   

   Warehouse Store  Watch Shop  Water Park  Weight Loss Center  Wine Bar  \
0              0.0         0.0         0.0                 0.0       0.0   
1              0.0         0.0         0.0                 0.0       0.0   
2              0.0         0.0         0.0                 0.0       0.0   
3              0.0         0.0         0.0                 0.0       0.0   
4              0.0         0.0         0.0                 0.0       0.0   

   Wine Shop  Wings Joint  Women's Store  
0        0.0          0.0            0.0  
1        0.0          0.0            0.0  
2        0.0          0.0            0.0  
3        0.0          0.0            0.0  
4        0.0          0.0            0.0  

[5 rows x 322 columns]

In [46]:
los_angeles_grouped.shape

(241, 322)

In [47]:
num_top_venues = 5
for hood in los_angeles_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = los_angeles_grouped[los_angeles_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Acton----
                           venue  freq
0     Construction & Landscaping   1.0
1          Performing Arts Venue   0.0
2                           Park   0.0
3  Paper / Office Supplies Store   0.0
4              Other Repair Shop   0.0


----Adams-Normandie----
              venue  freq
0  Sushi Restaurant  0.33
1              Park  0.11
2        Playground  0.11
3     Grocery Store  0.11
4       Gas Station  0.11


----Agoura Hills----
                  venue  freq
0  Fast Food Restaurant  0.13
1                Bakery  0.07
2    Chinese Restaurant  0.07
3     Indian Restaurant  0.07
4        Breakfast Spot  0.07


----Agua Dulce----
                           venue  freq
0                        Airport   1.0
1                    Yoga Studio   0.0
2                         Notary   0.0
3                           Park   0.0
4  Paper / Office Supplies Store   0.0


----Alhambra----
                 venue  freq
0    Convenience Store  0.31
1  Sporting Goods Shop  0.08
2     

In [48]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [77]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
la_neighborhoods_sorted = pd.DataFrame(columns=columns)
la_neighborhoods_sorted['Neighborhood'] = los_angeles_grouped['Neighborhood']

for ind in np.arange(los_angeles_grouped.shape[0]):
    la_neighborhoods_sorted.iloc[ind, 1:] = return_most_common_venues(los_angeles_grouped.iloc[ind, :], num_top_venues)

la_neighborhoods_sorted.head()

Neighborhood       1st Most Common Venue 2nd Most Common Venue  \
0            Acton  Construction & Landscaping         Women's Store   
1  Adams-Normandie            Sushi Restaurant           Gas Station   
2     Agoura Hills        Fast Food Restaurant    Chinese Restaurant   
3       Agua Dulce                     Airport         Women's Store   
4         Alhambra           Convenience Store   Sporting Goods Shop   

  3rd Most Common Venue 4th Most Common Venue  5th Most Common Venue  \
0                  Farm    English Restaurant  Entertainment Service   
1                  Park            Playground              Locksmith   
2        Breakfast Spot     Indian Restaurant                 Bakery   
3                  Farm    English Restaurant  Entertainment Service   
4           Video Store    Mexican Restaurant         Breakfast Spot   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0  Ethiopian Restaurant         Event Service            Eye Doctor   
1         Grocery Store            Taco Place         Women's Store   
2                  Café                   Gym      Sushi Restaurant   
3  Ethiopian Restaurant         Event Service            Eye Doctor   
4  Fast Food Restaurant             Pet Store        Hardware Store   

  9th Most Common Venue 10th Most Common Venue  
0           Fabric Shop     Falafel Restaurant  
1         Event Service     English Restaurant  
2           Gas Station                 Lounge  
3           Fabric Shop     Falafel Restaurant  
4            Bagel Shop            Pizza Place

In [72]:
kclusters = 15
laclustering = los_angeles_grouped.drop('Neighborhood', 1)
# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(laclustering)
# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10]

array([ 7,  1,  1, 14,  0,  9,  3,  1,  1,  1], dtype=int32)

In [78]:
# add clustering labels
la_neighborhoods_sorted.insert(0, 'Cluster Labels', kmeans.labels_)
la_merged = neighborhoods
# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
la_merged = la_merged.join(la_neighborhoods_sorted.set_index('Neighborhood'), on='Neighborhood')
la_merged.head() # check the last columns!

Neighborhood            Longitude            Latitude  Cluster Labels  \
0            Acton  -118.16981019229348  34.497355239240846             7.0   
1  Adams-Normandie  -118.30020800000011  34.031461499124156             1.0   
2     Agoura Hills  -118.75988450000015  34.146736499122795             1.0   
3       Agua Dulce   -118.3171036690717  34.504926999796837            14.0   
4         Alhambra  -118.13651200000021  34.085538999123571             0.0   

        1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0  Construction & Landscaping         Women's Store                  Farm   
1            Sushi Restaurant           Gas Station                  Park   
2        Fast Food Restaurant    Chinese Restaurant        Breakfast Spot   
3                     Airport         Women's Store                  Farm   
4           Convenience Store   Sporting Goods Shop           Video Store   

  4th Most Common Venue  5th Most Common Venue 6th Most Common Venue  \
0    English Restaurant  Entertainment Service  Ethiopian Restaurant   
1            Playground              Locksmith         Grocery Store   
2     Indian Restaurant                 Bakery                  Café   
3    English Restaurant  Entertainment Service  Ethiopian Restaurant   
4    Mexican Restaurant         Breakfast Spot  Fast Food Restaurant   

  7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0         Event Service            Eye Doctor           Fabric Shop   
1            Taco Place         Women's Store         Event Service   
2                   Gym      Sushi Restaurant           Gas Station   
3         Event Service            Eye Doctor           Fabric Shop   
4             Pet Store        Hardware Store            Bagel Shop   

  10th Most Common Venue  
0     Falafel Restaurant  
1     English Restaurant  
2                 Lounge  
3     Falafel Restaurant  
4            Pizza Place

In [79]:
la_merged = la_merged.fillna(0)
la_merged['Cluster Labels'] = la_merged['Cluster Labels'].astype(np.int64)
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(la_merged['Latitude'], la_merged['Longitude'], la_merged['Neighborhood'], la_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [float(lat), float(lon)],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

In [80]:
la_merged.loc[la_merged['Cluster Labels'] == 0, la_merged.columns[[1] + list(range(5, la_merged.shape[1]))]]

Longitude      2nd Most Common Venue   3rd Most Common Venue  \
4    -118.13651200000021        Sporting Goods Shop             Video Store   
8     -117.9223952817848                          0                       0   
12   -118.30465647554277                          0                       0   
14   -118.32733223477572        American Restaurant      Mexican Restaurant   
15    -118.0012614768012       Fast Food Restaurant           Grocery Store   
19    -117.9751905000002         Italian Restaurant      Mexican Restaurant   
21   -118.12903150000017             Sandwich Place                Pharmacy   
22    -118.1499360000002        Sporting Goods Shop      Seafood Restaurant   
31   -118.21078750000007                     Bakery                    Pool   
36   -118.60151949999999                 Sports Bar            Liquor Store   
37   -118.24800950000022                          0                       0   
39   -118.45476897959236                          0                       0   
41   -118.63334936248927                          0                       0   
46   -118.62695325711778                          0                       0   
51   -117.71876950000001                          0                       0   
54    -118.2218104930422           Department Store             Pizza Place   
66   -117.95762350000011                          0                       0   
67   -118.20484450000001              Grocery Store             Music Venue   
68   -118.19598900000028                  Pet Store              Food Truck   
74   -118.07396450000032              Grocery Store       Fish & Chips Shop   
78          -118.1762975                Pizza Place    Brazilian Restaurant   
86   -118.26860650000008                      Motel             Gas Station   
89   -118.26148999301319                 Donut Shop             Pizza Place   
93   -118.30898400000005                 Food Stand            Liquor Store   
95   -118.29715634932774       Fast Food Restaurant              Skate Park   
98   -118.37905350000011                          0                       0   
101          -118.307658              Grocery Store  Thrift / Vintage Store   
103  -118.22051147887569           Asian Restaurant      Mexican Restaurant   
105  -118.07219200000006               Liquor Store     Fried Chicken Joint   
106  -118.34605599620926       Fast Food Restaurant       Convenience Store   
..                   ...                        ...                     ...   
183   -117.7699420000001                   Pharmacy      Mexican Restaurant   
185  -118.21863657473801               Liquor Store             Gas Station   
190  -118.35453350000017                          0                       0   
192  -118.42312500000006              Grocery Store              Donut Shop   
193  -118.70332725511963                          0                       0   
196  -117.89399129564529          Mobile Phone Shop      Mexican Restaurant   
199  -118.43598850000011               Burger Joint          Ice Cream Shop   
202  -118.09900700000023         Mexican Restaurant     Japanese Restaurant   
205  -118.49231650000006                          0                       0   
207  -117.95659832063463                          0                       0   
209  -118.32881067473647                          0                       0   
214  -117.82609234720164                          0                       0   
215  -117.92981254313324                          0                       0   
220  -118.08974290054033       Fast Food Restaurant   Vietnamese Restaurant   
222   -118.0280174858583                          0                       0   
223  -118.45205006616247         Mexican Restaurant                    Food   
224  -118.58387954676732                          0                       0   
234  -118.28843919025437                          0                       0   
236  -118.45466051366508                          0              

In [81]:
la_merged.loc[la_merged['Cluster Labels'] == 1, la_merged.columns[[1] + list(range(5, la_merged.shape[1]))]]

Longitude         2nd Most Common Venue  \
1    -118.30020800000011                   Gas Station   
2    -118.75988450000015            Chinese Restaurant   
6    -118.08010100000017            Chinese Restaurant   
9    -118.03041899311202                           Gym   
10          -118.4307575                   Video Store   
11          -118.3234085                    Donut Shop   
13   -118.26237347966236                           Bar   
16   -117.91246849999999                  Liquor Store   
17   -118.29035750000011                Clothing Store   
20   -118.45841550000007              Football Stadium   
25   -118.42326299999999                 Women's Store   
26           -118.376102                     Juice Bar   
27   -118.32521100000017                           Spa   
28          -118.3049585                          Café   
29   -118.39953400000016                Breakfast Spot   
32   -117.96857400000013                    Food Court   
34   -118.27535153247453                  Liquor Store   
35   -118.67163173539396                    Food Court   
38            -118.36891                          Café   
43   -118.41508300000005                 Movie Theater   
47   -118.31320800000006      Mediterranean Restaurant   
48   -118.41146150000006          Other Great Outdoors   
49   -118.23700649999989         Vietnamese Restaurant   
50   -117.89216399278311                        Bakery   
52   -118.52796900000007                   Coffee Shop   
53    -118.1573625000002                  Home Service   
56   -118.00722099646512                     Multiplex   
57   -117.88392199287892                   Pizza Place   
58   -118.18557250000019          Fast Food Restaurant   
59   -118.40902450000016                   Coffee Shop   
..                   ...                           ...   
208  -118.49476193960714                       Dog Run   
210          -118.441514           American Restaurant   
212  -118.16716650000018                Sandwich Place   
213         -118.2677075               Thai Restaurant   
217  -118.19381199294568                           Gym   
219  -118.15623650000006                        Bakery   
221  -117.90184498555011                 Women's Store   
225  -118.38747919966544            Mexican Restaurant   
227   -118.3843172128845        Furniture / Home Store   
228  -118.54458900000007             Hotpot Restaurant   
229  -118.05489926123286                   Bus Station   
230  -118.35795100000007                Breakfast Spot   
231  -118.60842021657524                    Restaurant   
232  -118.35278600000026  Theme Park Ride / Attraction   
233  -118.35171088115038            Mexican Restaurant   
237  -118.28280750000005                    Food Truck   
238  -117.93079549289092                  Burger Joint   
239   -118.4145170000001     Middle Eastern Restaurant   
240  -118.39620200000005              Asian Restaurant   
243         -118.4639545                   Pizza Place   
244  -118.46284569710343                         Canal   
251  -117.86012600000016                          Farm   
252  -118.39829259016352                        Bakery   
256  -118.36961300000019                        Bakery   
258  -118.82291200000014                 Boat or Ferry   
259         -118.4307445            Persian Restaurant   
263  -118.07001998587738                  Burger Joint   
264  -118.44047994677246                   Coffee Shop   
269  -118.31990900000005                Cosmetics Shop   
271  -118.61521650000006          Arts & Entertainment   

         3rd Most Common Venue          4th Most Common Venue  \
1                         Park                     Playground   
2               Breakfast Spot              Indian Restaurant   
6              Bubble Tea Shop          Vietnamese Restaurant   
9             Football Stadium                 Gymnastics Gym   
10           Convenience Store                         Bakery   
11   Latin American Restaura

In [82]:
la_merged.loc[la_merged['Cluster Labels'] == 2, la_merged.columns[[1] + list(range(5, la_merged.shape[1]))]]

Longitude 2nd Most Common Venue 3rd Most Common Venue  \
91    -118.2449270000002         Women's Store    Falafel Restaurant   
161  -118.21900650000001                Garden                  Park   
191  -118.35980750108605         Women's Store    Falafel Restaurant   
211  -118.04883650000005          Gourmet Shop     Other Repair Shop   
235            -118.2818                 Trail         Women's Store   
241  -118.77026791367985         Women's Store    Falafel Restaurant   

    4th Most Common Venue 5th Most Common Venue  6th Most Common Venue  \
91    Empanada Restaurant    English Restaurant  Entertainment Service   
161         Women's Store     Electronics Store    Empanada Restaurant   
191   Empanada Restaurant    English Restaurant  Entertainment Service   
211           Coffee Shop            Eye Doctor    Empanada Restaurant   
235           Fabric Shop   Empanada Restaurant     English Restaurant   
241   Empanada Restaurant    English Restaurant  Entertainment Service   

     7th Most Common Venue  8th Most Common Venue 9th Most Common Venue  \
91    Ethiopian Restaurant          Event Service            Eye Doctor   
161     English Restaurant  Entertainment Service  Ethiopian Restaurant   
191   Ethiopian Restaurant          Event Service            Eye Doctor   
211     English Restaurant  Entertainment Service  Ethiopian Restaurant   
235  Entertainment Service   Ethiopian Restaurant         Event Service   
241   Ethiopian Restaurant          Event Service            Eye Doctor   

    10th Most Common Venue  
91             Fabric Shop  
161          Event Service  
191            Fabric Shop  
211          Event Service  
235             Eye Doctor  
241            Fabric Shop

In [83]:
la_merged.loc[la_merged['Cluster Labels'] == 3, la_merged.columns[[1] + list(range(5, la_merged.shape[1]))]]

Longitude   2nd Most Common Venue 3rd Most Common Venue  \
7    -118.13623898201556                    Food          Home Service   
23   -118.41421658346756            Home Service            Playground   
24   -118.17916600000018           Women's Store    Falafel Restaurant   
85   -118.24192048907494          Clothing Store         Grocery Store   
171  -118.41890850000007          Breakfast Spot         Women's Store   
247  -118.20343200000011  Furniture / Home Store    Falafel Restaurant   
255  -118.63539850000006      Italian Restaurant               Theater   

    4th Most Common Venue 5th Most Common Venue  6th Most Common Venue  \
7                Pharmacy            Campground          Women's Store   
23     Falafel Restaurant   Empanada Restaurant     English Restaurant   
24    Empanada Restaurant    English Restaurant  Entertainment Service   
85     Alternative Healer           Music Venue     Falafel Restaurant   
171                  Farm    English Restaurant  Entertainment Service   
247   Empanada Restaurant    English Restaurant  Entertainment Service   
255         Grocery Store                  Park      Martial Arts Dojo   

     7th Most Common Venue  8th Most Common Venue 9th Most Common Venue  \
7       English Restaurant  Entertainment Service  Ethiopian Restaurant   
23   Entertainment Service   Ethiopian Restaurant         Event Service   
24    Ethiopian Restaurant          Event Service            Eye Doctor   
85      English Restaurant  Entertainment Service  Ethiopian Restaurant   
171   Ethiopian Restaurant          Event Service            Eye Doctor   
247   Ethiopian Restaurant          Event Service            Eye Doctor   
255          Event Service            Flea Market   Empanada Restaurant   

    10th Most Common Venue  
7            Event Service  
23              Eye Doctor  
24             Fabric Shop  
85           Event Service  
171            Fabric Shop  
247            Fabric Shop  
255     English Restaurant

In [84]:
la_merged.loc[la_merged['Cluster Labels'] == 4, la_merged.columns[[1] + list(range(5, la_merged.shape[1]))]]

Longitude 2nd Most Common Venue 3rd Most Common Venue  \
92   -117.84178450000023              Mountain      Department Store   
135   -117.7745250000001           Music Store    Falafel Restaurant   
155  -118.45716250000007                Church         Women's Store   
160  -118.19427049999999                  Lake         Women's Store   
226  -118.31202250000007                  Park      Department Store   
248  -118.45725850000031            Food Truck         Garden Center   
262  -117.83509899236435     Other Repair Shop         Deli / Bodega   
266  -118.06118827749373                  Lake         Women's Store   

    4th Most Common Venue 5th Most Common Venue  6th Most Common Venue  \
92                   Farm    English Restaurant  Entertainment Service   
135   Empanada Restaurant    English Restaurant  Entertainment Service   
155   Empanada Restaurant    English Restaurant  Entertainment Service   
160   Empanada Restaurant    English Restaurant  Entertainment Service   
226                  Farm    English Restaurant  Entertainment Service   
248    Falafel Restaurant    English Restaurant  Entertainment Service   
262   Empanada Restaurant    English Restaurant  Entertainment Service   
266   Empanada Restaurant    English Restaurant  Entertainment Service   

    7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
92   Ethiopian Restaurant         Event Service            Eye Doctor   
135  Ethiopian Restaurant         Event Service            Eye Doctor   
155  Ethiopian Restaurant         Event Service            Eye Doctor   
160  Ethiopian Restaurant         Event Service            Eye Doctor   
226  Ethiopian Restaurant         Event Service            Eye Doctor   
248  Ethiopian Restaurant         Event Service            Eye Doctor   
262  Ethiopian Restaurant         Event Service            Eye Doctor   
266  Ethiopian Restaurant         Event Service            Eye Doctor   

    10th Most Common Venue  
92             Fabric Shop  
135            Fabric Shop  
155            Fabric Shop  
160            Fabric Shop  
226            Fabric Shop  
248            Fabric Shop  
262            Fabric Shop  
266            Fabric Shop

In [85]:
la_merged.loc[la_merged['Cluster Labels'] == 5, la_merged.columns[[1] + list(range(5, la_merged.shape[1]))]]

Longitude 2nd Most Common Venue 3rd Most Common Venue  \
108  -118.65705600000015         Women's Store    Falafel Restaurant   
125  -117.95562783562224         Women's Store    Falafel Restaurant   
156  -118.13161942411986      Business Service         Women's Store   

    4th Most Common Venue 5th Most Common Venue  6th Most Common Venue  \
108   Empanada Restaurant    English Restaurant  Entertainment Service   
125   Empanada Restaurant    English Restaurant  Entertainment Service   
156                  Farm    English Restaurant  Entertainment Service   

    7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
108  Ethiopian Restaurant         Event Service            Eye Doctor   
125  Ethiopian Restaurant         Event Service            Eye Doctor   
156  Ethiopian Restaurant         Event Service            Eye Doctor   

    10th Most Common Venue  
108            Fabric Shop  
125            Fabric Shop  
156            Fabric Shop

In [86]:
la_merged.loc[la_merged['Cluster Labels'] == 6, la_merged.columns[[1] + list(range(5, la_merged.shape[1]))]]

Longitude 2nd Most Common Venue 3rd Most Common Venue  \
168  -118.02918914224676         Women's Store     Electronics Store   

    4th Most Common Venue  5th Most Common Venue 6th Most Common Venue  \
168    English Restaurant  Entertainment Service  Ethiopian Restaurant   

    7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
168         Event Service            Eye Doctor           Fabric Shop   

    10th Most Common Venue  
168     Falafel Restaurant

In [87]:
la_merged.loc[la_merged['Cluster Labels'] == 7, la_merged.columns[[1] + list(range(5, la_merged.shape[1]))]]

Longitude 2nd Most Common Venue 3rd Most Common Venue  \
0    -118.16981019229348         Women's Store                  Farm   
94   -118.50767050010745         Women's Store                  Farm   
104  -118.66021608592953         Women's Store                  Farm   
128  -117.83603921804665         Women's Store                  Farm   

    4th Most Common Venue  5th Most Common Venue 6th Most Common Venue  \
0      English Restaurant  Entertainment Service  Ethiopian Restaurant   
94     English Restaurant  Entertainment Service  Ethiopian Restaurant   
104    English Restaurant  Entertainment Service  Ethiopian Restaurant   
128    English Restaurant  Entertainment Service  Ethiopian Restaurant   

    7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0           Event Service            Eye Doctor           Fabric Shop   
94          Event Service            Eye Doctor           Fabric Shop   
104         Event Service            Eye Doctor           Fabric Shop   
128         Event Service            Eye Doctor           Fabric Shop   

    10th Most Common Venue  
0       Falafel Restaurant  
94      Falafel Restaurant  
104     Falafel Restaurant  
128     Falafel Restaurant

In [88]:
la_merged.loc[la_merged['Cluster Labels'] == 8, la_merged.columns[[1] + list(range(5, la_merged.shape[1]))]]

Longitude 2nd Most Common Venue 3rd Most Common Venue  \
123  -118.37410748609757         Women's Store                  Farm   
174  -118.10809351531806            Food Court   Empanada Restaurant   
198  -117.81326850000013                Lawyer         Women's Store   

    4th Most Common Venue  5th Most Common Venue  6th Most Common Venue  \
123    English Restaurant  Entertainment Service   Ethiopian Restaurant   
174    English Restaurant  Entertainment Service   Ethiopian Restaurant   
198                  Farm     English Restaurant  Entertainment Service   

    7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
123         Event Service            Eye Doctor           Fabric Shop   
174         Event Service            Eye Doctor           Fabric Shop   
198  Ethiopian Restaurant         Event Service            Eye Doctor   

    10th Most Common Venue  
123     Falafel Restaurant  
174     Falafel Restaurant  
198            Fabric Shop

In [89]:
la_merged.loc[la_merged['Cluster Labels'] == 9, la_merged.columns[[1] + list(range(5, la_merged.shape[1]))]]

Longitude 2nd Most Common Venue      3rd Most Common Venue  \
5    -118.33515598608159          Home Service       Gym / Fitness Center   
18   -118.35774600000005          Home Service                       Park   
30   -118.39398199999999                  Park                 Food Truck   
42   -118.24721250000012            Donut Shop         Mexican Restaurant   
45   -117.85751998453503           Video Store           Business Service   
55   -118.22415649999999    Mexican Restaurant             Discount Store   
72    -118.1596383526979                  Park  Latin American Restaurant   
81   -118.29880750000001         Moving Target             Shop & Service   
96   -117.97508050764819        Scenic Lookout         Weight Loss Center   
99   -118.29516400000006    Spanish Restaurant               Home Service   
112  -118.33541000000008                  Park                Bus Station   
115   -118.2140149952952          Burger Joint        Fried Chicken Joint   
119  -118.32205850000005   Fried Chicken Joint               Home Service   
146  -118.20282648482208     Convenience Store                   Pharmacy   
173  -118.44521550000007       Automotive Shop         Mexican Restaurant   
184  -118.54872950000006                Lawyer                 Water Park   
216  -118.04853200000019     Convenience Store                       Park   
267  -118.25231247908229        Breakfast Spot              Grocery Store   

    4th Most Common Venue 5th Most Common Venue  6th Most Common Venue  \
5                    Park         Women's Store            Fabric Shop   
18          Women's Store           Fabric Shop    Empanada Restaurant   
30       Business Service              Boutique          Women's Store   
42   Kitchen Supply Store           Fabric Shop    Empanada Restaurant   
45          Moving Target          Dessert Shop    Empanada Restaurant   
55                 Bakery     Fish & Chips Shop            Fabric Shop   
72            Coffee Shop      Sushi Restaurant   Fast Food Restaurant   
81       Stationery Store                  Park           Liquor Store   
96            Bus Station           Fabric Shop    Empanada Restaurant   
99       Pedestrian Plaza    Mexican Restaurant            Fabric Shop   
112                 Trail           Fabric Shop    Empanada Restaurant   
115                  Farm    English Restaurant  Entertainment Service   
119     Convenience Store    Mexican Restaurant         Sandwich Place   
146                  Park         Women's Store             Eye Doctor   
173          Skating Rink         Women's Store            Fabric Shop   
184             Locksmith           Fabric Shop    Empanada Restaurant   
216         Women's Store    Falafel Restaurant    Empanada Restaurant   
267     Convenience Store    Falafel Restaurant    Empanada Restaurant   

     7th Most Common Venue            8th Most Common Venue  \
5      Empanada Restaurant               English Restaurant   
18      English Restaurant            Entertainment Service   
30                    Farm            Entertainment Service   
42      English Restaurant            Entertainment Service   
45   Entertainment Service             Ethiopian Restaurant   
55     Empanada Restaurant               English Restaurant   
72      Light Rail Station  Southern / Soul Food Restaurant   
81             Fabric Shop            Entertainment Service   
96      English Restaurant            Entertainment Service   
99     Empanada Restaurant               English Restaurant   
112     English Restaurant            Entertainment Service   
115   Ethiopian Restaurant                    Event Service   
119             Taco Place              Empanada Restaurant   
146      Electronics Store              Empanada Restaurant   
173     English Restaurant            Entertainment Service   
184     English Restaurant            Entertainment Service   
216     English Restaurant            Entertainment Service   
267     En